### ISSUE TITLE
[Feature Request] Modernize CUDA support in GPU package

### ISSUE THREAD

**Issue by akohlmey** [lammps/lammps#2355]
The CUDA support in the GPU package code in lib/gpu is starting to show its age.

On one hand, there are places with defines and workaround to support very old and now obsolete CUDA versions.
On the other hand, compiling the code with CUDA 11 gives warnings about deprecated APIs.

Thus it seems to be prudent to do the following:
- set a minimal CUDA toolkit version required from looking at the description of major feature changes and also at the GPU package code, something like version 4.0 or even 6.0 seems like a good choice
- remove all code only used with a CUDA toolkit predating the minimal version
- identify the deprecated APIs and implement replacements

In addition, somebody familiar with recent features in the CUDA toolkit may consider having a look at what could be possible improvements of the GPU kernels. This could be done in lockstep with modernizing OpenCL support and has to check that compatibility with the HIP backend is not broken, so this is a bit of a tall order.

**Comment by akohlmey** [comment]
The APIs flagged as deprecated with CUDA 11 are used in nvd_texture.h and are:
```
CUresult cuTexRefSetAddress_v2(size_t*, CUtexref, CUdeviceptr, size_t)
CUresult cuTexRefSetFormat(CUtexref, CUarray_format, int)

```

**Comment by maximtrp** [comment]
@akohlmey 
Would it be more reasonable to just get rid of dropped compute capabilities (like `sm_30` and `sm_32` in CUDA 11) and leave deprecated ones (`sm_35`, `sm_37` and `sm_50`) as is? Dropping deprecated CCs ahead of CUDA itself seems to be too strict and narrows `lammps` compatibility.

**Comment by akohlmey** [comment]
Are you sure you are not confusing "compute capability" with "CUDA toolkit
version"?

--
Dr. Axel Kohlmeyer akohlmey@gmail.com http://goo.gl/1wk0
College of Science & Technology, Temple University, Philadelphia PA, USA
International Centre for Theoretical Physics, Trieste, Italy

**Comment by maximtrp** [comment]
@akohlmey Oh, sorry! 🤦 You were talking about CUDA toolkit versions, not CCs. Just wanted to warn against dropping support for too many compute capabilities and not so old GPUs.

**Comment by akohlmey** [comment]
We'll have to revisit this topic in a while. A lot of issues should be resolved now that PR #2603 is included.
